### PR TITLE
#1219 main finance page

### LINKED ISSUE #1219: Finance - Main Reimbursement Dashboard (Finance View)
### Description

The finance team should have its own view of the reimbursement dashboard. 

### Acceptance Criteria

1. The two [FT] options should be grayed out unless viewed by the finance team
5. Clicking the "Pending Advisor List" button should open the "Pending Advisor List" popup
6. Clicking the "Generate All Receipts" button should open the "Generate All Receipts" popup

### Proposed Solution

^

### Mocks

image
epic #493

### PR DESCRIPTION
**Pull request by RChandler234**
## Changes

Had to add getAllReimbursement and getCurrentUser Reimbursement stuff to the backend, otherwise just added the api/hook stuff, had to do smthn weird to check if the user is on the finance team bc that env var is only used in the backend rn (lmk what you think ab that), and followed the mock, just pulled the tables out of the pageblocks
### Finance View
image
image
image
image

### Non-Finance View
image
image
image
image

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] No `yarn.lock` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #1219  (issue #)

### CODE REVIEW

**Review by zmar234 — COMMENTED** [review]
Looks Great! Small thing to consider, I wonder if Total Owed and Total Received should be flipped around so it's kind of read "amount received out of amount owed".

**Comment by RChandler234** [comment]
image
added recipients for all views bc I think it makes sense

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+} from '@mui/material';
+import { useState } from 'react';
+import { useAllReimbursements, useCurrentUserReimbursements } from '../../hooks/finance.hooks';
+import ErrorPage from '../ErrorPage';
+import LoadingIndicator from '../../components/LoadingIndicator';
+import { Reimbursement, ReimbursementRequest, User } from 'shared';
+import { useCurrentUser } from '../../hooks/users.hooks';
+import { fullNamePipe } from '../../utils/pipes';
+
+const NERProgressBar = styled(LinearProgress)(({ theme }) => ({
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:28]: put this in a component

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+  [`&.${linearProgressClasses.colorPrimary}`]: {
+    backgroundColor: theme.palette.grey[theme.palette.mode === 'light' ? 200 : 800]
+  },
+  [`& .${linearProgressClasses.bar}`]: {
+    borderRadius: 5,
+    backgroundColor: '#ef4345'
+  }
+}));
+
+const createRefundData = (date: Date, amount: number, recipient: User) => {
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:40]: put this in a utils file

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+interface RefundTableProps {
+  currentUserRequests: ReimbursementRequest[];
+  allRequests?: ReimbursementRequest[];
+}
+
+const Refunds = ({ currentUserRequests, allRequests }: RefundTableProps) => {
+  const [value, setValue] = useState(0);
+  const user = useCurrentUser();
+
+  const { data, isLoading, isError, error } = useCurrentUserReimbursements();
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:53]: descriptive names pls

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  const { data: allData, isLoading: allIsLoading, isError: allIsError, error: allError } = useAllReimbursements();
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:54]: more descriptive pls

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+  if (user.isFinance && allIsError) return <ErrorPage message={allError?.message} />;
+  if (isError) return <ErrorPage message={error?.message} />;
+  if ((user.isFinance && (allIsLoading || !allData)) || isLoading || !data) return <LoadingIndicator />;
+
+  const handleChange = (event: React.SyntheticEvent, newValue: number) => {
+    setValue(newValue);
+  };
+
+  const rows = (allData && value === 1 ? allData : data).map((row: Reimbursement) =>
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:65]: bro alldata and data is so confusing

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    createRefundData(row.dateCreated, row.amount, row.userSubmitted)
+  );
+  const totalReceived = (allData && value === 1 ? allData : data).reduce(
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:68]: make a constant for allData && value === 1 ? allData : data

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+  const totalOwed = (allRequests && value === 1 ? allRequests : currentUserRequests).reduce(
+    (accumulator: number, currentVal: ReimbursementRequest) => accumulator + currentVal.totalCost,
+    0
+  );
+  const percentRefunded = (totalReceived / totalOwed) * 100;
+
+  return (
+    <Box sx={{ bgcolor: theme.palette.background.paper, width: '100%', borderRadius: '8px 8px 8px 8px', boxShadow: 1 }}>
+      <AppBar sx={{ borderRadius: '8px 8px 0 0' }} position="static">
+        <Tabs value={value} onChange={handleChange} indicatorColor="secondary" textColor="inherit" variant="fullWidth">
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:81]: I am still in very much so am in favor of making a tab container component and then passing in a { tabData:  { label: string, value: number }[], onChange: (value) => void } object to standardize the tabs

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+      <Box
+        sx={{
+          backgroundColor: theme.palette.background.paper,
+          width: '100%',
+          padding: '10px',
+          borderRadius: '0 0 8px 8px'
+        }}
+      >
+        <Box sx={{ padding: '30px' }}>
+          <Box sx={{ display: 'flex', flexDirection: 'horizontal', justifyContent: 'space-between', paddingX: '30px' }}>
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:101]: why are we not using grids?

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            <Stack sx={{ alignItems: 'center' }}>
+              <Typography textAlign="center" sx={{ fontWeight: 700 }}>
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:103]: make this a component you use it multiple times

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
lets gooo. looks great just some cleanliness stuff

File: src/frontend/src/pages/FinancePage/FinancePage.tsx [hunk 10]
```diff
@@ -3,51 +3,97 @@
...
+          <ListItemIcon>
+            <ReceiptIcon fontSize="small" />
+          </ListItemIcon>
+          Generate All Receipts
+        </MenuItem>
+      </Menu>
+    </>
+  );
   return (
     <div>
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/FinancePage.tsx:87]: arbitrary

File: src/frontend/src/pages/FinancePage/FinancePage.tsx [hunk 11]
```diff
@@ -3,51 +3,97 @@
...
-            type="file"
-            accept="image/*"
-          />
-          <Typography>Reimbursement Request Id</Typography>
-          <TextField onChange={(e) => setReimbursementRequestId(e.target.value)}></TextField>
-          <Button type="submit"> Submit</Button>
-        </Stack>
-      </form>
+      <PageTitle title={'Reimbursement Requests'} previousPages={[]} actionButton={financeActionsDropdown} />
+      <Box sx={{ display: 'flex', flexDirection: 'horizontal' }}>
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/FinancePage.tsx:89]: cause grids is overkill for what we're doing here

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+  return { date, amount, recipient };
+};
+
+interface RefundTableProps {
+  currentUserRequests: ReimbursementRequest[];
+  allRequests?: ReimbursementRequest[];
+}
+
+const Refunds = ({ currentUserRequests, allRequests }: RefundTableProps) => {
+  const [value, setValue] = useState(0);
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/RefundsSection.tsx:50]: this is the tab value, I'll change the name

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+  );
+  const totalOwed = (allRequests && value === 1 ? allRequests : currentUserRequests).reduce(
+    (accumulator: number, currentVal: ReimbursementRequest) => accumulator + currentVal.totalCost,
+    0
+  );
+  const percentRefunded = (totalReceived / totalOwed) * 100;
+
+  return (
+    <Box sx={{ bgcolor: theme.palette.background.paper, width: '100%', borderRadius: '8px 8px 8px 8px', boxShadow: 1 }}>
+      <AppBar sx={{ borderRadius: '8px 8px 0 0' }} position="static">
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/RefundsSection.tsx:80]: just a component, that's how it has this different look for the tabs, I ripped it from the mui docs

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx (hunk 8 above)
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/RefundsSection.tsx:101]: again I don't think what we're making here is a grid, tbh we have probably been using grids more then we need to

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx (hunk 9 above)
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/RefundsSection.tsx:103]: sure

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+              </Typography>
+              <Typography> {`$${totalOwed}`}</Typography>
+            </Stack>
+          </Box>
+          <NERProgressBar sx={{ margin: '20px' }} variant="determinate" value={percentRefunded} />
+          <TableContainer component={Paper}>
+            <Table aria-label="simple table">
+              <TableHead>
+                <TableRow>
+                  <TableCell align="center" sx={{ fontSize: '16px', fontWeight: 600 }}>
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/RefundsSection.tsx:120]: ok

File: src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+  dateSubmitted: Date,
+  status: ReimbursementStatusType,
+  submitter: User,
+  saboId?: number,
+  dateDelivered?: Date
+) => {
+  return { saboId, amount, dateSubmitted, status, dateDelivered, submitter };
+};
+
+interface ReimbursementRequestTableProps {
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx:32]: nah, the point of these is to type the props being passed in, what if we wanna change the props being passed in later

File: src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+  Tabs,
+  useTheme
+} from '@mui/material';
+import { useState } from 'react';
+import { ReimbursementRequest, ReimbursementStatusType, User } from 'shared';
+import { getCurrentReimbursementStatus } from '../../utils/finance.utils';
+import { useCurrentUser } from '../../hooks/users.hooks';
+import { fullNamePipe } from '../../utils/pipes';
+
+const createRequestData = (
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx:21]: good idea

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx (hunk 7 above)
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/RefundsSection.tsx:81]: I can do it for these two and see if it actually helps, this one would be separate from the ones on all the overview/detail pages though

File: src/frontend/src/pages/FinancePage/FinancePage.tsx [hunk 17]
```diff
@@ -3,51 +3,97 @@
...
+
+  const handleClick = (event: React.MouseEvent<HTMLElement>) => {
+    setAnchorEl(event.currentTarget);
+  };
+
+  const handleDropdownClose = () => {
+    setAnchorEl(null);
   };
 
+  const financeActionsDropdown = (
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/FinancePage.tsx:48]: @Peyton-McKee which PR?

**Comment by RChandler234** [comment]
image

**Comment by RChandler234** [comment]
image

File: src/backend/src/transformers/auth-user.transformer.ts [hunk 18]
```diff
@@ -12,7 +12,11 @@ const authenticatedUserTransformer = (user: Prisma.UserGetPayload<typeof authUse
     role: user.role,
     defaultTheme: user.userSettings?.defaultTheme,
     teamAsLeadId: user.teamAsLead?.teamId,
-    favoritedProjectsId: user.favoriteProjects.map((project) => project.projectId)
+    favoritedProjectsId: user.favoriteProjects.map((project) => project.projectId),
+    isFinance:
+      !!process.env.FINANCE_TEAM_ID &&
+      (user.teams.map((team: Team) => team.teamId).includes(process.env.FINANCE_TEAM_ID) ||
```
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/backend/src/transformers/auth-user.transformer.ts]: Oh right thats seans pr i think but something to think about

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+    0
+  );
+  const totalOwed = displayedReimbursementRequests.reduce(
+    (accumulator: number, currentVal: ReimbursementRequest) => accumulator + currentVal.totalCost,
+    0
+  );
+  const percentRefunded = (totalReceived / totalOwed) * 100;
+
+  const tabs = [{ label: 'My Refunds', value: 0 }];
+  if (user.isFinance) tabs.push({ label: 'All Club Refunds', value: 1 });
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:85]: nicee

File: src/frontend/src/pages/FinancePage/RefundsSection.tsx [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+
+  return (
+    <Box sx={{ bgcolor: theme.palette.background.paper, width: '100%', borderRadius: '8px 8px 8px 8px', boxShadow: 1 }}>
+      <FinanceTabs tabValue={tabValue} setTabValue={setTabValue} tabs={tabs} />
+      <Box
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/RefundsSection.tsx:90]: cant these next two boxes be combined?

File: src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx:10]: we would make a new prop and if were just adding stuff we would just extend it, right? like these are identical props rn

File: src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+              <ColumnHeader title="Date Submitted" />
+              <ColumnHeader title="Date Delivered" />
+              <ColumnHeader title="Status" />
+            </TableRow>
+          </TableHead>
+          <TableBody>
+            {rows.map((row, index) => (
+              <TableRow key={`$${row.amount}-${index}`} sx={{ '&:last-child td, &:last-child th': { border: 0 } }}>
+                {tabValue === 1 && <TableCell align="center">{fullNamePipe(row.submitter)}</TableCell>}
+                <TableCell align="center">{row.saboId != null ? row.saboId : '-----'}</TableCell>
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx:52]: I have something similar here, I think we should make an undefined pipe and just pass in the saboId into it and it returns this string

File: src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+                <TableCell align="center">{row.amount}</TableCell>
+                <TableCell align="center">{datePipe(row.dateSubmitted)}</TableCell>
+                <TableCell align="center">{!!row.dateDelivered ? datePipe(row.dateDelivered) : '-----'}</TableCell>
```
Comment (Peyton-McKee) [inline src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx:55]: you could also use it here

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
looking good heres some more comments

File: src/backend/src/utils/reimbursement-requests.utils.ts [hunk 23]
```diff
@@ -215,6 +216,14 @@ export const validateUserIsPartOfFinanceTeam = async (user: UserWithTeam) => {
   }
 };
 
+export const isAuthUserOnFinanceTeam = (user: Prisma.UserGetPayload<typeof authUserQueryArgs>) => {
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/backend/src/utils/reimbursement-requests.utils.ts:219]: that's kinda weird if I'm understanding you correctly. Like return a boolean based on whether it throws an error? I think just having a separate method is just simpler

File: src/frontend/src/hooks/finance.hooks.ts [hunk 24]
```diff
@@ -15,3 +22,45 @@ export const useUploadSingleReceipt = (id: string) => {
     return data;
   });
 };
+
+/**
+ * Custom React Hook to get the reimbursement requests for the current user
+ *
+ */
+export const useCurrentUserReimbursementRequests = () => {
+  return useQuery<ReimbursementRequest[], Error>(['reimbursement request', 'user'], async () => {
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/frontend/src/hooks/finance.hooks.ts:31]: oh like they should all just invalidate each other?

File: src/backend/src/utils/reimbursement-requests.utils.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/backend/src/utils/reimbursement-requests.utils.ts:219]: oh I see what you mean with Sean's PR now, let me get his wrapped up then I'll do a little moving code around

File: src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx (hunk 21 above)
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx:52]: good idea

File: src/frontend/src/hooks/finance.hooks.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/frontend/src/hooks/finance.hooks.ts:31]: I think so? Idrk exactly how this works

File: src/backend/src/prisma-query-args/auth-user.query-args.ts [hunk 25]
```diff
@@ -1,10 +1,14 @@
 import { Prisma } from '@prisma/client';
+import teamQueryArgs from './teams.query-args';
 
 const authUserQueryArgs = Prisma.validator<Prisma.UserArgs>()({
   include: {
     userSettings: true,
     teamAsLead: true,
     favoriteProjects: true,
+    teams: {
```
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/backend/src/prisma-query-args/auth-user.query-args.ts:9]: I can't do an async call of that method in the transformer, so I think I need to do this

**Comment by RChandler234** [comment]
it won't let me reply to 2 of your comments, but I think having two separate prop interfaces is good just to keep the files looking the same (for newer devs or whoever else), I know it's duplicate code, I think it's fine. For using your dropdown component I think we should just merge this, and the add it in when we get to the PR with that in it since there's added stuff scattered across these PRs

File: src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx (hunk 22 above)
Comment (RChandler234) [reply src/frontend/src/pages/FinancePage/ReimbursementRequestsSection.tsx:55]: doesn't really work here bc we have to transform the object using datePipe

File: src/backend/src/prisma-query-args/auth-user.query-args.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (Peyton-McKee) [reply src/backend/src/prisma-query-args/auth-user.query-args.ts]: really? why not its just a method

**Review by Peyton-McKee — APPROVED** [review]
lets get this merged🍔

File: src/backend/src/prisma-query-args/auth-user.query-args.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (RChandler234) [reply src/backend/src/prisma-query-args/auth-user.query-args.ts]: wait I think I'm dumb lemme try smthn, will merge this v soon regardless of if it works

[the remark this answers is not in the document]
Comment (RChandler234) [reply src/backend/src/prisma-query-args/auth-user.query-args.ts]: forgot to await the method, but making the auth user transformer async was causing a weird issue where the app was just stuck loading forever. We can probably find a cleaner way to do this in the future, but for now I think including all the teams and just checking the team id's against the environment variable is probably the easiest/ has the least change to break smthn with the auth user

**Pull request merged**